ouroboros-network-framework/test/Test/Simulation/Network/Snocket.hs

```diff
@@ -627,28 +638,208 @@ prop_generator_NonFailingBeararInfoScript (NonFailingBearerInfoScript s) = not (
+
+prop_connect_to_accepting_socket :: BearerInfoScript -> Property
+prop_connect_to_accepting_socket (BearerInfoScript script) =
+    prop_verify_connect_to_simulation sim
+  where
+    sim :: forall s . IOSim s (Either SomeException ())
+    sim =
+      withSnocket nullTracer (toBearerInfo <$> script) $ \snocket ->
+        withAsync
+          (runServer (TestAddress (0 :: Int)) snocket (close snocket) acceptOne)
+          $ \serverAsync -> do
+            res <- runClient (TestAddress 1) (TestAddress 0)
+                            snocket (close snocket)
+            _ <- wait serverAsync
+            return res
+
+prop_connect_and_not_close :: BearerInfoScript -> Property
+prop_connect_and_not_close (BearerInfoScript script) =
+    prop_verify_connect_to_simulation sim
+  where
+    sim :: forall s . IOSim s (Either SomeException ())
+    sim =
+      withSnocket nullTracer (toBearerInfo <$> script) (\snocket ->
+        withAsync
+          (runServer (TestAddress (0 :: Int)) snocket (\_ -> pure ()) acceptOne)
+          $ \serverAsync -> do
+            res <- runClient (TestAddress 1) (TestAddress 0)
+                            snocket (\_ -> pure ())
+            _ <- wait serverAsync
+            return res
+                                                       )
+      `catch` \(err :: SomeException) ->
+        -- Should error with NotReleasedListeningSockets
+        case (fromException err
+               :: Maybe (ResourceException (TestAddress Int))) of
+          Just _ ->
+            return (Right ())
+          Nothing ->
+            return (Left err)
+
+prop_connect_to_not_accepting_socket :: BearerInfoScript -> Property
+prop_connect_to_not_accepting_socket (BearerInfoScript script) =
+    prop_verify_connect_to_simulation sim
+  where
+    sim :: forall s . IOSim s (Either SomeException ())
+    sim =
+      withSnocket nullTracer (toBearerInfo <$> script) $ \snocket ->
+        withAsync (runServer (TestAddress (0 :: Int)) snocket
+                             (close snocket) loop)
+          $ \_ -> do
+            res <- runClient (TestAddress 1) (TestAddress 0)
+                            snocket (close snocket)
+            case res of
+              -- Should timeout
+              Left _ -> return (Right ())
+              Right _ -> return (Left (toException UnexpectedOutcome))
+
+prop_connect_to_uninitialised_socket :: BearerInfoScript -> Property
+prop_connect_to_uninitialised_socket (BearerInfoScript script) =
+    prop_verify_connect_to_simulation sim
+  where
+    sim :: forall s . IOSim s (Either SomeException ())
+    sim =
+      withSnocket nullTracer (toBearerInfo <$> script) $ \snocket -> do
+        res <- runClient (TestAddress (1 :: Int)) (TestAddress 0)
+                        snocket (close snocket)
+        case res of
+          -- Should complain about no such listening socket
+          Left _ -> return (Right ())
+          Right _ -> return (Left (toException UnexpectedOutcome))
```

Comment:
Just for completeness sake, could we also add a test which connects to a socket that is bound but is not a listening?